lib/opte/src/engine/nat.rs

```diff
@@ -24,40 +24,146 @@ use alloc::string::ToString;
 use alloc::sync::Arc;
 use alloc::vec::Vec;
 use core::fmt;
+use core::hash::Hash;
 use core::marker::PhantomData;
+use crc32fast::Hasher;
 use opte_api::Direction;
 use opte_api::IpAddr;
 
-/// A mapping from a private to external IP address for NAT.
-#[derive(Debug, Clone, Copy)]
-pub struct Nat {
+/// A trait which allows a VPC implementation to specify how NAT actions
+/// can be re-verified after a rule change.
+///
+/// This is needed for outbound flows in particular, as the flow id and opaque
+/// action alone don't allow us to see the chosen external IpAddr. For the inbound
+/// case, the gateway layer can successfully rematch if needed but reusing this
+/// mechanism is the most sensible approach.
+pub trait VerifyAddr: alloc::fmt::Debug + Send + Sync {
+    fn is_addr_valid(&self, addr: &IpAddr) -> bool;
+}
+
+/// A mapping from a private to one of several external IP addresses for NAT.
+#[derive(Debug, Clone)]
+pub struct OutboundNat {
     priv_ip: IpAddr,
-    external_ip: IpAddr,
+    // TODO: possibly remove Vec on ephemeral IP.
+    external_ips: Vec<IpAddr>,
+
+    verifier: Arc<dyn VerifyAddr>,
+}
+
+impl OutboundNat {
+    /// Create a new NAT mapping from a private to public IP address.
+    pub fn new<T: ConcreteIpAddr>(
+        priv_ip: T,
+        external_ips: &[T],
+        verifier: Arc<impl VerifyAddr + 'static>,
+    ) -> Self {
+        let external_ips = external_ips.iter().copied().map(T::into).collect();
+        Self { priv_ip: priv_ip.into(), external_ips, verifier }
+    }
+}
+
+impl fmt::Display for OutboundNat {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{} <=> ", self.priv_ip)?;
+
+        // Sadly, there is no `formatter::Fmt::display_list()`.
+        if self.external_ips.len() > 1 {
+            write!(f, "{{")?;
+        }
+
+        for (i, ip) in self.external_ips.iter().enumerate() {
+            write!(f, "{ip}")?;
+            if i != self.external_ips.len() - 1 {
+                write!(f, ",")?;
+            }
+        }
```

Comment:
We already have `itertools` as a dep right? Could use [`Itertools::format`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.format):
```suggestion
        write!(f, "{}", self.external_ips.iter().format(","));
```